spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -0,0 +1,371 @@
+import org.apache.spark.sql.execution.datasources.parquet.ParquetTest
+import org.apache.spark.sql.functions.lit
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types._
+
+/**
+ * Suite covering the type widening table feature.
+ */
+class DeltaTypeWideningSuite
+  extends QueryTest
+    with ParquetTest
+    with DeltaDMLTestUtils
+    with DeltaSQLCommandTest
+    with DeltaTypeWideningTestMixin
+    with DeltaTypeWideningAlterTableTests
+    with DeltaTypeWideningTableFeatureTests
+
+/**
+ * Test mixin that enables type widening by default for all tests in the suite.
+ */
+trait DeltaTypeWideningTestMixin extends SharedSparkSession {
+  protected override def sparkConf: SparkConf = {
+    super.sparkConf.set(DeltaConfigs.ENABLE_TYPE_WIDENING.defaultTablePropertyKey, "true")
+  }
+
+  /** Enable (or disable) type widening for the table under the given path. */
+  protected def enableTypeWidening(tablePath: String, enabled: Boolean = true): Unit =
+    sql(s"ALTER TABLE delta.`$tablePath` " +
+          s"SET TBLPROPERTIES('${DeltaConfigs.ENABLE_TYPE_WIDENING.key}' = '${enabled.toString}')")
+}
+
+/**
+ * Trait collecting a subset of tests providing core coverage for type widening using ALTER TABLE
+ * CHANGE COLUMN TYPE.
+ */
+trait DeltaTypeWideningAlterTableTests extends QueryErrorsBase {
+  self: QueryTest with ParquetTest with DeltaTypeWideningTestMixin with DeltaDMLTestUtils =>
+
+  import testImplicits._
+
+  /**
+   * Represents the input of a type change test.
+   * @param fromType         The original type of the column 'value' in the test table.
+   * @param toType           The type to use when changing the type of column 'value'
+   * @param initialValues    The initial values to insert in column 'value after table creation,
+   *                         using type `fromType`
+   * @param additionalValues Additional values to insert after changing the type of the column
+   *                         'value' to `toType`.
+   */
+  case class TypeEvolutionTestCase(
+      fromType: DataType,
+      toType: DataType,
+      initialValues: Seq[String],
+      additionalValues: Seq[String] = Seq.empty) {
+    def initialValuesDF: DataFrame =
+      initialValues.toDF("value").select($"value".cast(fromType))
+
+    def additionalValuesDF: DataFrame =
+      additionalValues.toDF("value").select($"value".cast(toType))
+
+    def expectedResult: DataFrame =
+      initialValuesDF.union(additionalValuesDF).select($"value".cast(toType))
+  }
+
+  // Type changes that are supported by all Parquet readers. Byte, Short, Int are all stored as
+  // INT32 in parquet so these changes are guaranteed to be supported.
+  private val supportedTestCases: Seq[TypeEvolutionTestCase] = Seq(
+    TypeEvolutionTestCase(ByteType, ShortType,
+      Seq("1", "2", Byte.MinValue.toString),
+      Seq("4", "5", Int.MaxValue.toString)),
```

Comment:
`Short`?

The "after" cast would just return NULL, I guess. How about setting config ANSI_ENABLED for this test, to avoid using values that don't actually fit?

Reply:
Good catch, switched to running with ANSI enabled